tests/test_network_failure_duplicate_messages.py

```diff
@@ -0,0 +1,247 @@
+        assert 2 == mock_read_file.call_count, "Expected server to invoke read_file method twice"
+        assert response_1, "Expected server to respond to first message"
+        assert response_2, "Expected server to respond to second message"
+
+    def test_network_failure_experiment_2(self):
+        """
+        Network experiment 2: simulate duplicate of read request message.
+
+        Things to test:
+        - Server should invoke the read_file method only once for the first message.
+        - Server should use the message history to respond to the second message.
+
+        innovcation     : at most once
+        operation type  : idempotent
+
+        """
+        server_root_directory: Path = Path.cwd() / "tests" / "server"
+        server_file_system = ServerFileSystem(server_root_directory=server_root_directory)
+        mock_server_file_system = Mock(wraps=server_file_system)
+        server = Server(
+            server_ip_address=self.SERVER_IP_ADDRESS,
+            server_port_number=self.SERVER_PORT_NUMBER,
+            file_system=mock_server_file_system,
+            invocation_semantics=InvocationSemantics.AT_MOST_ONCE,
+        )
+        logger.info("SERVER START")
+        server_thread = threading.Thread(target=server.listen_for_messages)
+        server_thread.start()
+
+        fixed_uuid: UUID = uuid4()
+        logger.info("Sending the first message")
+        response_1 = send_message_and_wait_for_reply(
+            message=ReadFileRequest(request_id=fixed_uuid, filename="english_alphabets.txt"),
+            recipient_ip_address=self.SERVER_IP_ADDRESS,
+            recipient_port_number=self.SERVER_PORT_NUMBER,
+            max_attempts_to_send_message=3,
+            timeout_in_seconds=5,
+        )
+        time.sleep(2)
+        logger.info("Sending the second message")
+        response_2 = send_message_and_wait_for_reply(
+            message=ReadFileRequest(request_id=fixed_uuid, filename="english_alphabets.txt"),
+            recipient_ip_address=self.SERVER_IP_ADDRESS,
+            recipient_port_number=self.SERVER_PORT_NUMBER,
+            max_attempts_to_send_message=3,
+            timeout_in_seconds=3,
+        )
+        mock_read_file: Mock = mock_server_file_system.read_file
+        server.stop_listening()
+        server_thread.join()
+        mock_read_file.assert_called_once(), "Expected server to invoke read_file method only once"
+        assert response_1, "Expected server to respond to first message"
+        assert response_1 == response_2, "Expected server to use cached reply to respond to the second message"
+
+    def test_network_failure_experiment_5(self):
+        """
+        Network experiment 5: simulate duplicate of append request message.
+
+        Things to test:
+        - Server should invoke the read_file method only once for the first message.
+        - Server should use the message history to respond to the second message.
+
+        innovcation     : at most once
```

Comment:
I think you meant at-least-once?

Reply:
Yup. Make a mistake here. Corrected. Thanks.